File: CODE/funAckley.f90
!=========================================
! Ackley function
!
! n=2 (default)
!   
! molti minimi locali       
!
! valore ottimo della funzione obiettivo = 0.0
!
! punto ottimo = 0
!
!         -15.0 =< x(i) =< 30.0   
!====================================================    

SUBROUTINE FUNCT(N,X,F)

      IMPLICIT NONE

      INTEGER          :: N
      DOUBLE PRECISION :: X(N), F

      DOUBLE PRECISION, PARAMETER :: PI = 3.1415923D0

      DOUBLE PRECISION :: a,b,c,s1,s2
      INTEGER          :: I

      a = 20.d0
	  b = 0.2d0
	  c = 2.d0*pi

      s1 = 0.d0 
	  s2 = 0.d0

      DO I=1,N
         s1 = s1+x(i)*x(i)
         s2 = s2+dcos(c*x(i))
      ENDDO

      F = -a*dexp(-b*sqrt((1.d0/dble(n))*s1))-dexp((1.d0/dble(n))*s2)+a+dexp(1.d0)

      RETURN

END



subroutine grad(n,x,g)

implicit none

integer			::n,i
double precision::x(n), g(n)
double precision :: a,b,c,s1,s2
double precision, parameter :: pi = 3.1415923d0

a = 20.d0
b = 0.2d0
c = 2.d0*pi

s1 = 0.d0 
s2 = 0.d0

do i=1,n
 s1 = s1+x(i)*x(i)
 s2 = s2+dcos(c*x(i))
enddo

do i = 1,n
     
	g(i) = a*b*dexp(-b*sqrt((1.d0/dble(n))*s1)) * (1.d0/(dsqrt((1.d0/dble(n))*s1)))&
	 *x(i)*1.d0/dble(n) + dexp((1.d0/dble(n))*s2)*dsin(c*x(i))*c*(1.d0/dble(n))
	
enddo
return

end subroutine grad


SUBROUTINE DATA(N,L,U,FOTT)

      IMPLICIT NONE

      INTEGER          :: N
      DOUBLE PRECISION :: L(N),U(N),FOTT
      INTEGER          :: I

      DO I=1,N
         L(I) = -15.d0
         U(I) =  30.d0
      ENDDO

      FOTT = 0.D0

      RETURN

END

File: Python/funAckley.py
#=========================================
# Ackley function
#
# n=2 (default)
#   
# molti minimi locali       
#
# valore ottimo della funzione obiettivo = 0.0
#
# punto ottimo = 0
#
#         -15.0 =< x(i) =< 30.0   
#====================================================    

import numpy as np

def ackley_function(x):
    a = 20.0
    b = 0.2
    c = 2.0 * np.pi

    s1 = np.sum(x**2)
    s2 = np.sum(np.cos(c * x))

    f = -a * np.exp(-b * np.sqrt((1.0 / len(x)) * s1)) - np.exp((1.0 / len(x)) * s2) + a + np.exp(1.0)

    return f

def ackley_gradient(x):
    a = 20.0
    b = 0.2
    c = 2.0 * np.pi

    s1 = np.sum(x**2)
    s2 = np.sum(np.cos(c * x))

    g = np.zeros_like(x)

    for i in range(len(x)):
        g[i] = a * b * np.exp(-b * np.sqrt((1.0 / len(x)) * s1)) * (1.0 / np.sqrt((1.0 / len(x)) * s1)) * \
               x[i] * (1.0 / len(x)) + np.exp((1.0 / len(x)) * s2) * np.sin(c * x[i]) * c * (1.0 / len(x))

    return g

def initialize_bounds(n):
    lower_bound = np.full(n, -15.0)
    upper_bound = np.full(n, 30.0)
    return lower_bound, upper_bound

if __name__ == "__main__":
    # Example usage
    n_dim = 2
    x_example = np.zeros(n_dim)

    function_value = ackley_function(x_example)
    gradient_value = ackley_gradient(x_example)
    lower_bound, upper_bound = initialize_bounds(n_dim)

    print("Function value:", function_value)
    print("Gradient value:", gradient_value)
    print("Lower bounds:", lower_bound)
    print("Upper bounds:", upper_bound)
